prov/verbs/src/verbs_domain_xrc.c

```diff
@@ -0,0 +1,576 @@
+	struct ofi_rbnode *node;
+	int ret;
+
+	assert(ep->id);
+
+	fi_ibv_set_ini_conn_key(ep, &key);
+	node = ofi_rbmap_find(domain->xrc.ini_conn_rbmap, &key);
+	if (node) {
+		ini_conn = node->data;
+		ofi_atomic_inc32(&ini_conn->ref_cnt);
+		return ini_conn;
+	}
+
+	ini_conn = calloc(1, sizeof(*ini_conn));
+	if (!ini_conn) {
+		errno = FI_ENOMEM;
+		return NULL;
+	}
+
+	ini_conn->tgt_qpn = FI_IBV_NO_INI_TGT_QPNUM;
+	ini_conn->peer_addr = mem_dup(key.addr, ofi_sizeofaddr(key.addr));
+	if (!ini_conn->peer_addr) {
+		free(ini_conn);
+		errno = FI_ENOMEM;
+		return NULL;
+	}
+	ini_conn->tx_cq = container_of(ep->util_ep.tx_cq,
+				       struct fi_ibv_cq, util_cq);
+	dlist_init(&ini_conn->pending_list);
+	dlist_init(&ini_conn->active_list);
+	ofi_atomic_initialize32(&ini_conn->ref_cnt, 1);
+
+	ret = ofi_rbmap_insert(domain->xrc.ini_conn_rbmap,
+			       (void *) &key, (void *) ini_conn);
+	assert(ret != -FI_EALREADY);
+	if (ret) {
+		VERBS_WARN(FI_LOG_FABRIC, "INI QP RBTree insert failed %d\n",
+			   ret);
+		errno = -ret;
+		goto insert_err;
+	}
+	return ini_conn;
+
+insert_err:
+	free(ini_conn->peer_addr);
+	free(ini_conn);
+	return NULL;
+}
+
+/* Caller must hold domain:xrc:ini_mgmt_lock */
+void fi_ibv_put_shared_ini_conn(struct fi_ibv_ep *ep)
+{
+	struct fi_ibv_domain *domain = fi_ibv_msg_ep_to_domain(ep);
+	struct fi_ibv_ini_shared_conn *ini_conn = ep->ini_conn;
+	struct fi_ibv_ini_conn_key key;
+	struct ofi_rbnode *node;
+
+	if (!ini_conn)
+		return;
+
+	/* remove from pending or active connection list */
+	dlist_remove(&ep->ini_conn_entry);
+	ep->conn_state = FI_IBV_XRC_UNCONNECTED;
+	ep->ini_conn = NULL;
+	ep->ibv_qp = NULL;
+
+	/* Tear down physical INI/TGT when no longer being used */
+	if (!ofi_atomic_dec32(&ini_conn->ref_cnt)) {
+		if (ibv_destroy_qp(ini_conn->ini_qp))
+			VERBS_WARN(FI_LOG_FABRIC, "destroy of QP error %d\n",
+				   errno);
+
+		fi_ibv_set_ini_conn_key(ep, &key);
+		node = ofi_rbmap_find(domain->xrc.ini_conn_rbmap, &key);
+		assert(node);
+		ofi_rbmap_delete(domain->xrc.ini_conn_rbmap, node);
+		free(ini_conn->peer_addr);
+		free(ini_conn);
+		return;
+	}
+
+	fi_ibv_sched_ini_conn(ini_conn);
+}
+
+/* Caller must hold domain:xrc:ini_mgmt_lock */
+void fi_ibv_add_pending_ini_conn(struct fi_ibv_ep *ep, int reciprocal,
+				 void *conn_param, size_t conn_paramlen)
+{
+	ep->conn_setup->pending_recip = reciprocal;
+	ep->conn_setup->pending_paramlen = MIN(conn_paramlen,
+				sizeof(ep->conn_setup->pending_param));
+	memcpy(ep->conn_setup->pending_param, conn_param,
+	       ep->conn_setup->pending_paramlen);
+	dlist_insert_tail(&ep->ini_conn_entry, &ep->ini_conn->pending_list);
+}
+
+static void fi_ibv_create_shutdown_event(struct fi_ibv_ep *ep)
+{
+	struct fi_eq_cm_entry entry = {
+		.fid = &ep->util_ep.ep_fid.fid,
+	};
+
+	fi_ibv_eq_write_event(ep->eq, FI_SHUTDOWN, &entry, sizeof(entry));
+}
+
+static int fi_ibv_process_ini_conn(struct fi_ibv_ep *ep,int reciprocal,
+				   void *param, size_t paramlen);
```

Comment:
Please place all forward function declarations at the top of the file

Reply:
Updated.